### PR TITLE
CHD support (credits to @SleepyMan and @siddhartha77 too)

### PR DESCRIPTION
**Pull request by rtissera**
This PR tries to resolve remaining issues on https://github.com/PCSX2/pcsx2/pull/4045 and @siddhartha77 work in order to finally get CHD support in.

### CODE REVIEW

**Comment by liberodark** [comment]
Thank you for this pr.

**Comment by rtissera** [comment]
(hopefully) fixed the remaining clang issue upstream with https://github.com/rtissera/libchdr/commit/0b6d6635eaab4b35928e704aa9f33cba99d9297e
Can someone relaunch the CI or should I do a dummy commit ?

**Comment by rtissera** [comment]
sorry for all the noise on this PR, I had issues reproducing locally some of your CI scenarios.
Let me know if you can squash and merge or want me to manually redo a "clean" PR from scratch.

**Comment by lightningterror** [comment]
Would be nice to properly rebase the branch with latest master.

Also name the commits properly with prefixes included.

**Comment by rtissera** [comment]
@lightningterror let me know if you need another rebase or if it's clean enough to be merged.

File: pcsx2/CDVD/ChdFileReader.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
+#include "PrecompiledHeader.h"
```
Comment (lightningterror) [inline pcsx2/CDVD/ChdFileReader.cpp:1]: Missing copyright header and need to run clang format on the file.

File: pcsx2/CDVD/ChdFileReader.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#pragma once
```
Comment (lightningterror) [inline pcsx2/CDVD/ChdFileReader.h:1]: Same.

File: pcsx2/CDVD/CompressedFileReader.cpp [hunk 3]
```diff
@@ -16,12 +16,16 @@
...
 #include "AsyncFileReader.h"
 #include "CompressedFileReader.h"
+#include "ChdFileReader.h"
 #include "CsoFileReader.h"
 #include "GzippedFileReader.h"
 
 // CompressedFileReader factory.
 AsyncFileReader* CompressedFileReader::GetNewReader(const wxString& fileName)
 {
+	if (ChdFileReader::CanHandle(fileName)) {
```
Comment (lightningterror) [inline pcsx2/CDVD/CompressedFileReader.cpp:26]: Formatting.

File: pcsx2/CDVD/ChdFileReader.cpp (hunk 1 above)
Comment (rtissera) [reply pcsx2/CDVD/ChdFileReader.cpp:16]: Fixed in last commit.

File: pcsx2/CDVD/ChdFileReader.h (hunk 2 above)
Comment (rtissera) [reply pcsx2/CDVD/ChdFileReader.h:16]: Fixed in last commit.

File: pcsx2/CDVD/CompressedFileReader.cpp (hunk 3 above)
Comment (rtissera) [reply pcsx2/CDVD/CompressedFileReader.cpp]: Fixed in last commit.

File: pcsx2/CDVD/ChdFileReader.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2010  PCSX2 Dev Team
```
Comment (lightningterror) [inline pcsx2/CDVD/ChdFileReader.cpp:2]: 2021 instead of 2010

File: pcsx2/CDVD/ChdFileReader.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2010  PCSX2 Dev Team
```
Comment (lightningterror) [inline pcsx2/CDVD/ChdFileReader.h:2]: Ditto.

Comment (rtissera) [reply pcsx2/CDVD/ChdFileReader.h]: indeed !

File: pcsx2/CDVD/ChdFileReader.cpp (hunk 4 above)
Comment (rtissera) [reply pcsx2/CDVD/ChdFileReader.cpp]: indeed !

**Comment by lightningterror** [comment]
I'd squash the last commit with the previous 2 ones.

**Comment by rtissera** [comment]
@lightningterror Sure no problem for the squash, can you do it or should I ?

**Comment by lightningterror** [comment]
You can do it.

**Comment by rtissera** [comment]
@lightningterror done, good to go ?

**Review by lightningterror — APPROVED** [review]

**Review by refractionpcsx2 — APPROVED** [review]

**Review by tadanokojin — APPROVED** [review]

**Review by RedDevilus — APPROVED** [review]

**Review by GovanifY — APPROVED** [review]

**Review by jackun — APPROVED** [review]

**Review by F0bes — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

**Comment by Florin9doi** [comment]
Closes #2584, closes #4045

**Comment by refractionpcsx2** [comment]
@rtissera sorry can you fixt he conflicts? Thanks

**Comment by rtissera** [comment]
@refractionpcsx2 sure no problem will fix later today

**Comment by rtissera** [comment]
@refractionpcsx2 rebased on master with fixed conflicts

**Comment by refractionpcsx2** [comment]
Awesome, thanks, now we can merge :)

**Comment by matoro** [comment]
Hello, is there going to be an option to use the system version of this new library instead of the submodule one, like yaml-cpp?

**Comment by tony971** [comment]
CHD files are not listed as options under "Boot ISO"

**Comment by refractionpcsx2** [comment]
> CHD files are not listed as options under "Boot ISO"

Boot ISO is the same thing.

**Comment by tony971** [comment]
> > CHD files are not listed as options under "Boot ISO"
> 
> Boot ISO is the same thing. We can rename it "Boot Image" if it really bothers you.

I didn't communicate effectively enough. If you go to "Boot ISO", it won't find chd files. It's not one of the available options.

**Comment by refractionpcsx2** [comment]
If you're using a buildbot build, they are out of date, for some reason it isn't building, so CHD isn't available in those builds yet. we are working on it.

**Comment by tony971** [comment]
The latest buildbot is bc477e1. That looks to be the head.

**Comment by refractionpcsx2** [comment]
> The latest buildbot is [bc477e1](https://github.com/PCSX2/pcsx2/commit/bc477e1ce10a54c279b681a8d6b99f34d5d0b8fb). That looks to be the head.

no, the last buildbot build is dated 2021-03-15 13:16:20, which is from before the CHD stuff.

You can't go off what's on the PCSX2 page, that lists every commit regardless, you need to go actually on to the buildbot to see the correct build.

**Comment by RedDevilus** [comment]
Orphis buildbots (1.7-1149):
image

Actual latest from github (1.7-1154):
image

**Comment by tony971** [comment]
Yep, PEBCAK. Sorry for wasting your time.

**Comment by refractionpcsx2** [comment]
No problems, you're not the first who's been confused.

**Pull request merged**